### DISCUSSION TITLE
conflicting declaration  khronos_uint64_t

### DISCUSSION THREAD

**Discussion by vassilcom** [ocornut/imgui#4514] [New Users Build/Link/Run issues ONLY!]
Hello,

After Commit: 459de65477423360176447e79df2f3a785b71f3d on  Aug 19 2021 by Rokas Kupstys  ( OpenGL3: Embed our own minimal GL loader based on gl3w and use it in SDL/GLFW examples.) I'm facing an compiling error on Linux (glad + sdl2/glfw):

imgui_impl_opengl3_loader.h:57:32: error: conflicting declaration ‘typedef long long int khronos_int64_t’

libs/glad/include/KHR/khrplatform.h:152:33: note: previous declaration as ‘typedef int64_t khronos_int64_t’

Does anyone have any experience in resolving this issue maybe?

Regards,
Vasili

**Comment by ocornut** [comment]
@vassilcom Can you expect how your compilation unit (1 source file) can use BOTH Glad and `imgui_impl_opengl3_loader` ?
This is strange..

**Reply by ocornut** [reply]
Just to clarify you should never have to include our imgui_impl_opengl3_loader.h file

**Comment by vassilcom** [comment]
That's my wrapper for OpenGL  "OpenGLWrapper.h":
`#pragma once

#ifdef WM_SDL
	#if defined(__EMSCRIPTEN__)
		#include <GLES2/gl2.h>
	#elif __ANDROID__
		#include <SDL_opengles2.h>
	#elif __linux__
		#define IMGUI_IMPL_OPENGL_LOADER_GLAD
		#include <glad/glad.h>
	#elif _WIN32
		#define IMGUI_IMPL_OPENGL_LOADER_GLAD
		#include <glad/glad.h>
	#endif
#elif WM_GLFW
	#if __linux__
		#define IMGUI_IMPL_OPENGL_LOADER_GLAD
		#include <glad/glad.h>
	#elif _WIN32
		#define IMGUI_IMPL_OPENGL_LOADER_GLAD
		#include <glad/glad.h>
	#endif
#endif`

and how I include headers:

`#include "imgui.h"

#ifdef WM_GLFW
    #include "imgui_impl_glfw.h"
#elif WM_SDL
    #include "imgui_impl_sdl.h"
#endif

#include "imgui_impl_opengl3.h"

#ifdef WM_GLFW
	#include <GLFW/glfw3.h>
#elif WM_SDL
	#include <SDL.h>
#endif

#include "OpenGLWrapper.h"`

I'll try further to find the reason, and will post it here if succeeded

**Reply by ocornut** [reply]
That code you posted doesn't use `imgui_impl_opengl3_loader.h` at all.

**Reply by vassilcom** [reply]
That's true! I don't use imgui_impl_opengl3_loader.h at all, and that's why it's bit weird . This file was introduced in Commit: 459de65 and it is exactly where error appeared. I guess this header is called  from imgui_impl_opengl3.cpp:

```
// GL includes
#if defined(IMGUI_IMPL_OPENGL_ES2)
#include <GLES2/gl2.h>
#if defined(__EMSCRIPTEN__)
#ifndef GL_GLEXT_PROTOTYPES
#define GL_GLEXT_PROTOTYPES
#endif
#include <GLES2/gl2ext.h>
#endif
#elif defined(IMGUI_IMPL_OPENGL_ES3)
#if (defined(__APPLE__) && (TARGET_OS_IOS || TARGET_OS_TV))
#include <OpenGLES/ES3/gl.h>    // Use GL ES 3
#else
#include <GLES3/gl3.h>          // Use GL ES 3
#endif
#else
// About Desktop OpenGL function loaders:
//  Modern desktop OpenGL doesn't have a standard portable header file to load OpenGL function pointers.
//  Helper libraries are often used for this purpose! Here we are using our own minimal custom loader based on gl3w.
//  You may use another loader/header of your choice (glext, glLoadGen, etc.), or chose to manually implement your own.
//  If you happen to be developing a new feature for this backend, you may want to get a unstripped version of
//  imgui_impl_opengl3_loader.h from https://github.com/dearimgui/gl3w_stripped/releases/ and use that temporarily.
//  When done, changes using new APIs should be accompanied by regenerated stripped loader version (instructions in
//  gl3w_stripped README.rst).
#define IMGL3W_IMPL 1
#include "imgui_impl_opengl3_loader.h"
#endif
```

but I'm trying not to touch imgui source files in any way..

**Reply by vassilcom** [reply]
if i revert imgui_impl_opengl3.cpp file, that is,  revert changes in it made by 459de65 commit, everything works perfectly again.

**Reply by rokups** [reply]
Could you share a preprocessed version of `imgui_impl_opengl3.cpp` that your project compiles?

**Reply by ocornut** [reply]
@vassilcom TL;DR; you have a file somewhere that include both glad.h and imgui_impl_opengl3_loader.h there's no reason ever to do so as by definition they are providing the same service.

We added a `#ifndef __khrplatform_h_` header guard to make this case maybe less problematic but unsure if other problems would manifest:
https://github.com/ocornut/imgui/commit/755bf985d2c8ca29a83e7a2900907a3992910edb

**Comment by AidanSun05** [comment]
*(EDIT: This answer is based on the modifications from e3988a8, the commit after the one you mentioned. This commit amends the backend to allow you to tell Dear ImGui that you have your own custom loader, so the stripped one won't be included.)*

---

> `#define IMGUI_IMPL_OPENGL_LOADER_GLAD`

This macro is no longer checked by the OpenGL 3 backend, so that's why it's always including `imgui_impl_opengl3_loader.h`. Replace it with `IMGUI_IMPL_OPENGL_LOADER_CUSTOM` to disable the inclusion of the Dear ImGui loader. You can then include Glad as usual and the duplicate declarations should go away.

Here's the section in the OpenGL3 backend that includes the stripped loader:

https://github.com/ocornut/imgui/blob/e3988a84db0a3e9ab3e826aa1da3262e6e1ad25a/backends/imgui_impl_opengl3.cpp#L112-L122

The `#include` is not compiled when `IMGUI_IMPL_OPENGL_LOADER_CUSTOM` is defined.

**Reply by vassilcom** [reply]
thank you! IMGUI_IMPL_OPENGL_LOADER_CUSTOM solved my problem. 
this line: " #elif !defined(IMGUI_IMPL_OPENGL_LOADER_CUSTOM)" comes in the following commits, and that's why i missed it.

**Reply by ocornut** [reply]
I repeat, you should not be using `IMGUI_IMPL_OPENGL_LOADER_CUSTOM` for that purpose. I would rather have us actually understand what wrong you are doing in order to fix it at the source. You are not providing us the right information so far :(

**Reply by vassilcom** [reply]
i'm bit confused if frankly .. perhaps I'm lacking some experience with c++. to go stepbystep:

1. I have a project that uses glad: my_project.h :
#include "imgui_impl_glfw.h"
#include <GLFW/glfw3.h>
#define IMGUI_IMPL_OPENGL_LOADER_CUSTOM
#include <glad/glad.h>

2. in a different cpp i'm including imgui files to use the respective functions: my_imgui_build.cpp:
#include "imgui_impl_opengl3.cpp"
#include "imgui_impl_glfw.cpp"

3.  in imgui/backends/imgui_impl_opengl3.cpp we have:
#elif !defined(IMGUI_IMPL_OPENGL_LOADER_CUSTOM)
#include "imgui_impl_opengl3_loader.h"
#endif

so, I guess defining IMGUI_IMPL_OPENGL_LOADER_CUSTOM is the only way to prevent loading imgui_impl_opengl3_loader.h, or do i miss something?

**Reply by PathogenDavid** [reply]
As a general rule of thumb, you should not include `.cpp` files. You need to build `imgui_impl_opengl3.cpp` and `imgui_impl_glfw.cpp` independently and link them into your application.

**Comment by ocornut** [comment]
Aidan: they should not need to use the CUSTOM define thats for way more advanced use cases. Our backend use its own loader. User code can use the loader of their choice. 

It is still unconfirmed how/why user code need to include both their loader and ours in a same source file. One reason i can see is they are using unity/amalgamated builds?
